File: supabase/functions/_shared/crypto.ts
/**
 * AES-256-GCM decryption helpers for OAuth token storage.
 * Compatible with Python backend crypto.py format:
 *   base64( IV[12 bytes] + ciphertext + GCM_tag[16 bytes] )
 */

const IV_LENGTH = 12;
const KEY_LENGTH = 32;

function base64ToBytes(b64: string): Uint8Array {
  const binary = atob(b64);
  const bytes = new Uint8Array(binary.length);
  for (let i = 0; i < binary.length; i++) {
    bytes[i] = binary.charCodeAt(i);
  }
  return bytes;
}

async function getKey(): Promise<CryptoKey> {
  const raw = Deno.env.get("ENCRYPTION_KEY");
  if (!raw) {
    throw new Error("ENCRYPTION_KEY is not set");
  }
  const keyBytes = base64ToBytes(raw);
  if (keyBytes.length !== KEY_LENGTH) {
    throw new Error(
      `ENCRYPTION_KEY must be ${KEY_LENGTH} bytes after base64 decode, got ${keyBytes.length}`
    );
  }
  return crypto.subtle.importKey(
    "raw",
    keyBytes,
    { name: "AES-GCM" },
    false,
    ["decrypt"],
  );
}

/**
 * Decrypt an AES-256-GCM encrypted token produced by Python's encrypt_token().
 *
 * The encrypted string is base64-encoded: IV (12 bytes) || ciphertext || GCM tag (16 bytes).
 * Web Crypto API expects tag appended to ciphertext, which matches Python's AESGCM output.
 */
export async function decryptToken(encrypted: string): Promise<string> {
  const key = await getKey();
  const raw = base64ToBytes(encrypted);
  const iv = raw.slice(0, IV_LENGTH);
  const ciphertextWithTag = raw.slice(IV_LENGTH);

  const plaintext = await crypto.subtle.decrypt(
    { name: "AES-GCM", iv },
    key,
    ciphertextWithTag,
  );

  return new TextDecoder().decode(plaintext);
}

File: backend/app/utils/crypto.py
"""AES-256-GCM encryption helpers for OAuth token storage."""

import base64
import os

from cryptography.hazmat.primitives.ciphers.aead import AESGCM

from app.config import settings

_IV_LENGTH = 12  # 96-bit IV for AES-GCM
_KEY_LENGTH = 32  # 256-bit key


def _get_key() -> bytes:
    """Decode the encryption key from settings."""
    raw = settings.encryption_key
    if not raw:
        raise RuntimeError("ENCRYPTION_KEY is not set")
    key = base64.b64decode(raw)
    if len(key) != _KEY_LENGTH:
        raise RuntimeError(f"ENCRYPTION_KEY must be {_KEY_LENGTH} bytes after base64 decode")
    return key


def encrypt_token(plaintext: str) -> str:
    """Encrypt a plaintext token using AES-256-GCM.

    Returns a base64-encoded string containing IV + ciphertext + tag.
    """
    key = _get_key()
    iv = os.urandom(_IV_LENGTH)
    aesgcm = AESGCM(key)
    ciphertext = aesgcm.encrypt(iv, plaintext.encode("utf-8"), None)
    return base64.b64encode(iv + ciphertext).decode("ascii")


def decrypt_token(encrypted: str) -> str:
    """Decrypt an AES-256-GCM encrypted token.

    Expects the base64-encoded string produced by encrypt_token().
    """
    key = _get_key()
    raw = base64.b64decode(encrypted)
    iv = raw[:_IV_LENGTH]
    ciphertext = raw[_IV_LENGTH:]
    aesgcm = AESGCM(key)
    plaintext = aesgcm.decrypt(iv, ciphertext, None)
    return plaintext.decode("utf-8")
